core/consensus/babe/src/lib.rs

```diff
@@ -197,80 +211,104 @@ pub struct BabeParams<C, E, I, SO, SC> {
 	pub force_authoring: bool,
 
 	/// The source of timestamps for relative slots
-	pub time_source: BabeLink,
+	pub babe_link: BabeLink<B>,
 }
 
 /// Start the babe worker. The returned future should be run in a tokio runtime.
-pub fn start_babe<B, C, SC, E, I, SO, Error, H>(BabeParams {
-	config,
-	client,
+pub fn start_babe<B, C, SC, E, I, SO, Error>(BabeParams {
 	keystore,
+	client,
 	select_chain,
-	block_import,
 	env,
+	block_import,
 	sync_oracle,
 	inherent_data_providers,
 	force_authoring,
-	time_source,
-}: BabeParams<C, E, I, SO, SC>) -> Result<
+	babe_link,
+}: BabeParams<B, C, E, I, SO, SC>) -> Result<
 	impl futures01::Future<Item=(), Error=()>,
 	consensus_common::Error,
 > where
-	B: BlockT<Header=H>,
-	C: ProvideRuntimeApi + ProvideCache<B> + ProvideUncles<B> + Send + Sync + 'static,
+	B: BlockT<Hash=H256>,
+	C: ProvideRuntimeApi + ProvideCache<B> + ProvideUncles<B> + BlockchainEvents<B>
+		+ HeaderBackend<B> + Send + Sync + 'static,
 	C::Api: BabeApi<B>,
 	SC: SelectChain<B> + 'static,
 	E: Environment<B, Error=Error> + Send + Sync,
 	E::Proposer: Proposer<B, Error=Error>,
 	<E::Proposer as Proposer<B>>::Create: Unpin + Send + 'static,
-	H: Header<Hash=B::Hash>,
-	I: BlockImport<B> + Send + Sync + 'static,
+	I: BlockImport<B,Error=ConsensusError> + Send + Sync + 'static,
 	Error: std::error::Error + Send + From<::consensus_common::Error> + From<I::Error> + 'static,
 	SO: SyncOracle + Send + Sync + Clone,
 {
+	let config = babe_link.config;
 	let worker = BabeWorker {
 		client: client.clone(),
 		block_import: Arc::new(Mutex::new(block_import)),
 		env,
 		sync_oracle: sync_oracle.clone(),
 		force_authoring,
-		c: config.c(),
 		keystore,
+		epoch_changes: babe_link.epoch_changes.clone(),
+		config: config.clone(),
 	};
-	register_babe_inherent_data_provider(&inherent_data_providers, config.0.slot_duration())?;
+
+	register_babe_inherent_data_provider(&inherent_data_providers, config.slot_duration())?;
 	uncles::register_uncles_inherent_data_provider(
 		client.clone(),
 		select_chain.clone(),
 		&inherent_data_providers,
 	)?;
-	Ok(slots::start_slot_worker(
+
+	let epoch_changes = babe_link.epoch_changes.clone();
+	let pruning_task = client.finality_notification_stream()
```

Comment:
This will only be started by validators this way and it must be run by all full nodes. That's why it was returned with the import queue.

Reply:
Yeah, I realized that earlier today, which a bit too late. Still, I don't think that forcing the user to spawn a background task is friendly at all. 

I'd rather have us prune implicitly every time we add an epoch, based on the finalized block.